src/query-builder/SelectQueryBuilder.ts

```diff
@@ -1939,6 +1939,10 @@ export class SelectQueryBuilder<Entity extends ObjectLiteral>
             return undefined
         }
 
+        if (entitiesAndRaw.entities.length === 0) {
+            return undefined;
+        }
```

Comment:
In cases where skip/offset were not used, the lazy count can still be trusted when the result set is empty. So let's move this down and condition it on that:


```ts
        const hasSkip =
this.expressionMap.skip !== undefined &&
this.expressionMap.skip !== null &&
this.expressionMap.skip > 0
        const hasOffset =
this.expressionMap.offset !== undefined &&
this.expressionMap.offset !== null &&
this.expressionMap.offset > 0

        if (entitiesAndRaw.entities.length === 0 && (hasSkip || hasOffset)) {
            // when skip or offset were used and no results found, we need to execute a full count
            // (the given offset may have exceeded the actual number of rows)
            return undefined
        }
```